### PR TITLE
fix(ci): skip pkg.pr.new when no packages need publishing

### PR DESCRIPTION
**Pull request by f0rdream**
## Summary

- Skip `pkg.pr.new` on `changeset-release/*` branches and when the PR itself produces no publishable release, including no-changeset and empty-changeset cases.
- Preserve failures from malformed changesets, invalid Git references, and package resolution instead of treating every error as a successful skip.
- Once eligible, publish the complete merge-ref release plan so selected packages, dependency propagation, manifests, and build artifacts describe the same merged snapshot.

## Motivation

`changeset status` exits with code 1 when publishable packages change without a changeset. This made WIP and non-release changes fail CI, encouraging contributors to add empty changesets that unnecessarily trigger the repository-wide CODEOWNERS rule.

Catching every non-zero exit would hide real failures. At the same time, base-branch changesets should not make a PR with no release publish a preview. The workflow therefore separates the decision to publish from the contents of an eligible merge preview.

## Design

Let `M` be the merge base, `H` the PR head, `B` the current `main` tip, and `C` GitHub's synthetic merge ref.

```mermaid
flowchart TB
  M["M: merge base"] --> H["H: PR head"]
  M --> B["B: current main"]
  H --> C["C: merge ref"]
  B --> C

  H --> Gate["Eligibility plan from M..H"]
  Gate --> Decision{"PR has a publishable release?"}
  Decision -->|No| Skip["Skip preview"]
  Decision -->|Yes| Plan["Complete active release plan from C"]
  C --> Plan
  Plan --> Publish["Publish C manifests and artifacts"]
```

`M..H` is only an eligibility gate:

- no PR changeset -> skip;
- empty or non-publishable PR changeset -> skip;
- valid PR release -> continue.

After the gate passes, Changesets reads every active changeset in `C` and computes the complete release and dependency propagation plan. This intentionally includes active base-branch changesets because the preview represents the merged workspace. pkg.pr.new receives the resulting package directories explicitly; it does not calculate changed packages itself.

## Changes

| Area | Change |
| --- | --- |
| Eligibility | Exclude `changeset-release/*`; inspect `M..H` and skip when the PR itself has no publishable release. |
| Merge preview | Build artifacts from `C`, calculate its complete active Changesets plan, and publish package paths from the same workspace. |
| Stale PR support | Load the current helper from the merge-ref workspace while inspecting the PR-head worktree. |
| Failure safety | Keep malformed changesets, invalid refs, and package resolver failures fatal. |
| Argument safety | Pass package directories as newline-delimited values and quoted Bash arrays so path boundaries are preserved. |
| Coverage | Add focused changeset-file tests, register them in CI, and add the missing license header to the existing package resolver. |

## Validation

```text
TEST: changeset file tests; changeset repository validations
TEST: actionlint; ESLint; dprint; CSpell
TEST: merge-plan and shell edge cases; pnpm turbo build (57/57)
```

No package versions or public APIs change, so this pull request intentionally has no non-empty changeset.

## Related

- Release-branch example: [#287][release-pr]
- pkg.pr.new directory behavior: [Publish Packages][pkg-pr-new-docs]

[release-pr]: https://github.com/lynx-family/lynx-ui/pull/287
[pkg-pr-new-docs]: https://github.com/stackblitz-labs/pkg.pr.new#publish-packages

### CODE REVIEW

**Review by MoonfaceX — DISMISSED** [review]

**Review by Dugyu — DISMISSED** [review]
Not ready.
Request changes : The workflow reprocesses deferred-component changesets already on`main` but absent from the PR base, causing pkg.pr.new to publish unrelated packages.

File: .github/workflows/pkg-pr-new.yml [hunk 1]
```diff
@@ -57,31 +65,61 @@ jobs:
...
+            "$GITHUB_WORKSPACE/node_modules/.bin/changeset" status \
+              --since="$MERGE_BASE" \
+              --output .changeset-status.json
+          )
+          package_dirs=()
+          while IFS= read -r package_dir; do
+            package_dirs+=("$package_dir")
+          done < <(
+            node .github/scripts/list-changeset-packages.cjs \
+              "$PR_HEAD_DIR/.changeset-status.json"
```
[the remark this answers is not in the document]
Comment (Dugyu) [reply .github/workflows/pkg-pr-new.yml]: Thanks for checking this. Running package discovery from the merge-ref workspace is intentional here.

The PR-head worktree is used only to determine the release names introduced by the PR (`MERGE_BASE..PR_HEAD_SHA`). `list-changeset-packages.cjs` then maps those names to package paths in the merge-ref workspace, which is the tree actually published together with its extracted build artifacts and manifests.

The helper emits absolute `pkg.path` values, and `$PR_HEAD_DIR` is removed by the `EXIT` trap before the separate publish step runs. Running discovery from that worktree would therefore pass paths that no longer exist. If workspace definitions differ, the merge-ref tree is also the authoritative post-merge state we need to publish. No code change is needed for this finding.

File: .github/workflows/pkg-pr-new.yml [hunk 2]
```diff
@@ -54,34 +62,68 @@ jobs:
...
+          )
+          # Resolve package names in the merge-ref workspace. These absolute
+          # paths remain valid after the temporary PR-head worktree is removed.
+          package_dirs=()
+          while IFS= read -r package_dir; do
+            package_dirs+=("$package_dir")
+          done < <(
+            node .github/scripts/list-changeset-packages.cjs \
+              "$PR_HEAD_DIR/.changeset-status.json"
+          )
```
[the remark this answers is not in the document]
Comment (Dugyu) [reply .github/workflows/pkg-pr-new.yml]: Addressed in the latest commit [4035a58](https://github.com/lynx-family/lynx-ui/pull/294/commits/4035a587be3baa69e7c8ee6f28b995a6625345c6).

**Review by fzx2666-fz — DISMISSED** [review]

File: .github/workflows/pkg-pr-new.yml [hunk 3]
```diff
@@ -30,7 +35,10 @@ jobs:
     needs: build
     runs-on: ubuntu-latest
     timeout-minutes: 30
-    if: github.repository == 'lynx-family/lynx-ui'
+    if: >-
+      github.repository == 'lynx-family/lynx-ui' &&
+      !startsWith(github.head_ref, 'changeset-release/') &&
+      !startsWith(github.head_ref, 'release/')
```
Comment (fzx2666-fz) [inline .github/workflows/pkg-pr-new.yml]: release/ doesn't exist

**Review by Dugyu — DISMISSED** [review]
Request changes.

File: .github/workflows/pkg-pr-new.yml (hunk 3 above)
Comment (Dugyu) [reply .github/workflows/pkg-pr-new.yml]: fixed
